crates/ra_hir/src/ty/infer.rs

```diff
@@ -503,85 +503,71 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
 
     fn resolve_assoc_item(
         &mut self,
-        mut def: TypeNs,
+        def: TypeNs,
         path: &Path,
         remaining_index: usize,
         id: ExprOrPatId,
     ) -> Option<(ValueNs, Option<Substs>)> {
```

Comment:
Can we actually resolve more than once segment as assoc type? Something like this fails to compile: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ff72fdd4c27f8da5443153fd33b29d93

When I've touched this code with name res refactoring, I had a feeling that some parts here are dead code

Reply:
Ah, paths can be longer for functions:

```
let _: X::Item = X::Item::default();
```

Reply:
Yeah, that. I think I got rid of any loops, we still have to check whether there are more segments remaining though.